homeassistant/components/firmata/__init__.py

```diff
@@ -0,0 +1,142 @@
+    {
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_PIN): vol.Any(cv.positive_int, ANALOG_PIN_SCHEMA),
+        # will be analog mode in future too
+        vol.Required(CONF_PIN_MODE): vol.Any(PIN_MODE_INPUT, PIN_MODE_PULLUP),
+        vol.Optional(CONF_NEGATE_STATE, default=False): cv.boolean,
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+BOARD_CONFIG_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_SERIAL_PORT): cv.string,
+        vol.Optional(CONF_SERIAL_BAUD_RATE): cv.positive_int,
+        vol.Optional(CONF_ARDUINO_INSTANCE_ID): cv.positive_int,
+        vol.Optional(CONF_ARDUINO_WAIT): cv.positive_int,
+        vol.Optional(CONF_SLEEP_TUNE): vol.All(
+            vol.Coerce(float), vol.Range(min=0.0001)
+        ),
+        vol.Optional(CONF_SAMPLING_INTERVAL): cv.positive_int,
+        vol.Optional(CONF_SWITCHES): [SWITCH_SCHEMA],
+        vol.Optional(CONF_BINARY_SENSORS): [BINARY_SENSOR_SCHEMA],
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.All(cv.ensure_list, [BOARD_CONFIG_SCHEMA])}, extra=vol.ALLOW_EXTRA
+)
+
+
+async def async_setup(hass, config):
+    """Set up the Firmata domain."""
+    # Delete all current entries and re-add them to update them
+    if hass.config_entries.async_entries(DOMAIN) and DOMAIN in config:
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            await hass.config_entries.async_remove(entry.entry_id)
+    if DOMAIN in config:
+        for board in config[DOMAIN]:
+            firmata_config = deepcopy(board)
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN,
+                    context={"source": config_entries.SOURCE_IMPORT},
+                    data=firmata_config,
+                )
+            )
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up a Firmata board for a config entry."""
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
+
+    _LOGGER.debug(
+        "Setting up Firmata id %s, config %s", config_entry.entry_id, config_entry.data
+    )
+    if not config_entry.options:
+        hass.config_entries.async_update_entry(config_entry)
+
+    board = FirmataBoard(hass, config_entry)
+
+    if not await board.async_setup():
+        return False
+
+    hass.data[DOMAIN][board.name] = board
+
+    await board.async_update_device_registry()
+    if CONF_BINARY_SENSORS in config_entry.data:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, "binary_sensor")
+        )
+    if CONF_SWITCHES in config_entry.data:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, "switch")
+        )
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, board.async_reset)
+    return True
+
+
+async def async_remove_entry(hass, config_entry) -> None:
```

Comment:
The implementation looks like it's for unloading? 
```suggestion
async def async_unload_entry(hass, config_entry) -> None:
```

Removing should only be done for final cleanups when an entry is deleted.

Reply:
You also need to unload all platforms that you set up.

Reply:
OK, so I guess I'm a bit confused with the difference between these methods. This is what I have working now:
```python
async def async_unload_entry(hass, config_entry) -> None:
    """Shutdown and close a Firmata board for a config entry."""

    _LOGGER.info("Closing Firmata board %s", config_entry.data[CONF_NAME])
    if CONF_BINARY_SENSORS in config_entry.data:
        await hass.config_entries.async_forward_entry_unload(config_entry, "binary_sensor")
    if CONF_SWITCHES in config_entry.data:
        await hass.config_entries.async_forward_entry_unload(config_entry, "switch")
    await hass.data[DOMAIN].pop(config_entry.entry_id).async_reset()
    return True
```
I also added `async_added_to_hass()` and `async_will_remove_from_hass()` where needed in the platforms so that they work properly.

However, switching to this function seems to create an issue: if a user uses the "delete" button on the integrations page, then the integration deletes successfully and runs this code to safely close the board, but then when HA is restarted all of the entities have `_2` appended to their names even though the user thinks the integration and its entities were removed. What am I missing?

This is also not doing what I expect/want it to do: ideally this is called when HA is shutting down so that the board is closed and reset properly, but that doesn't seem to be happening (or maybe I misread the docs and this is not a possible feature).

I have fixed everything else, so a commit should come soon :smiley:

Reply:
When HA shuts down we send de `EVENT_HOMEASSISTANT_STOP` event, you should listen for that (and clean it up on unload)

Unload is called when your config entry is unloaded. This can happen for various reasons, but mainly because of deletion.

remove_entry is called when it is actually deleted. Use this callback if you need to tell ie a remote server to drop a webhook, because it is not needed anymore.

Reply:
When things are appended with `_2`, it is caused by unique IDs that are not matching .

Reply:
> When HA shuts down we send de EVENT_HOMEASSISTANT_STOP event, you should listen for that (and clean it up on unload)
> 
> Unload is called when your config entry is unloaded. This can happen for various reasons, but mainly because of deletion.

OK, so I will use `async_unload_entry` and listen for `EVENT_HOMEASSISTANT_STOP`.

> When things are appended with _2, it is caused by unique IDs that are not matching .

Ah, so that makes sense since I am now indexing my `entry_id`.